File: nqueen.py
'''
argv[1] : Board size
'''

import random
import sys
from operator import itemgetter
import time

start = 0

#Boardを出力
def board_print(gene, size):
    for i in gene:
        for j in range(size):
            if i == j:
                print("Q", end=" ")
            else:
                print(".", end=" ")
        print("")


#遺伝子生成
def makeIniGene(gene_num, size):
    ini_gene = []
    for cnt in range(0, gene_num):
        line = []
        for i in range(0, size):
            line.append(i)

        random.shuffle(line)
        ini_gene.append(line)

    return ini_gene


#評価関数
def calcFitness(gene, size):
    fitness = 0
    gVec = [(-1,-1), (-1,1), (1,-1), (1,1)]

    for i in range(size):
        for vec in gVec:
            for j in range(1, size):
                x = vec[1] * j + gene[i]
                y = vec[0] * j + i
                if x < 0 or x >= size or y < 0 or y >= size:
                    break
                
                if x == gene[y]:
                    fitness += 1
    
    return fitness


#順列表現と順序表現へ変換
def p_to_o(gene, size):
    converted = []
    num_list = []
    #数値が1からサイズ分まで順番に入っているリストを作成
    for num in range(size):
        num_list.append(num)
    
    #遺伝子変換
    for num in gene:
        converted.append(num_list.index(num))
        num_list.remove(num)

    return converted


#順序表現から順列表現へ変換
def o_to_p(gene, size):
    converted = []
    num_list = []
    
    for num in range(size):
        num_list.append(num)

    for num in gene:
        converted.append(num_list.pop(num))

    return converted


#交叉
#遺伝子のリストを渡すと交叉する
def cross (ini_gene, size):
    for i in range(int(len(ini_gene)/2)):
        n = i * 2
        num = random.randint(1,size-2)

        tmp = ini_gene[n][0:num]
        tmp.extend(ini_gene[n+1][num:])

        ini_gene[n+1] = ini_gene[n+1][0:num]
        ini_gene[n+1].extend(ini_gene[n][num:])
        
        ini_gene[n] = tmp


#突然変異
def mutation(gene_list, size):
    gene = random.randint(0,len(gene_list)-1)
    num1 = random.randint(0, size-1)
    num2 = random.randint(0, size-1)

    tmp = gene_list[gene][num1]
    gene_list[gene][num1] = gene_list[gene][num2]
    gene_list[gene][num2] = tmp


#適応度を基準にソートし，淘汰と増殖を行う
def gene_sort(gene_list, size, count):
    fit_gene = []
    for gene in gene_list:
        add_gene = []
        add_gene.append(gene[:])
        add_gene.append(calcFitness(gene, size))
        fit_gene.append(add_gene)

    fit_gene.sort(key=itemgetter(1))

    print(fit_gene[0][1])
    '''
    for i in range(len(gene_list)):
        print(fit_gene[i][0])
    print("")
    '''
    #board_print(fit_gene[0][0], size)
    
    #適応度が0ならプログラム終了
    if fit_gene[0][1] == 0:
        print("count {}".format(count))
        board_print(fit_gene[0][0], size)
        end = time.time()
        print("Time:{}[sec]".format(end - start))
        sys.exit(0)

    gene_list = []
    for gene in fit_gene:
        gene_list.append(gene.pop(0))

    #淘汰は2つ行う
    gene_list[-1] = gene_list[0]
    gene_list[1] = gene_list[0]

    return gene_list


#NQueen本体
def main(argv):
    #NxNのBoard，N個のQueen，のN
    size = int(argv[1])

    #１世代あたりの遺伝子の数
    gene_num = 10

    #初期集団の生成
    gene_list = makeIniGene(gene_num, size)

    #ループに入る
    n = 1
    while True:
        #遺伝子が評価され，淘汰され，増殖する
        gene_list = gene_sort(gene_list, size, n)

        #交叉する
        cross_gene = []
        for gene in gene_list:
            cross_gene.append(p_to_o(gene, size))

        cross(cross_gene, size)

        gene_list = []
        for gene in cross_gene:
            gene_list.append(o_to_p(gene, size))

        #突然変異する
        mutation(gene_list, size)

        n += 1


if __name__ == "__main__":
    start = time.time()
    main(sys.argv)
    

File: nqueen.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

clock_t start;

/* 構造体宣言 */
//遺伝子
typedef struct {
    int* gene;
    int* fitness;
}gene_struct;

typedef struct cell{
    int n;
    struct cell* next;
}list; 

/* 関数プロトタイプ宣言 */
void board_print(int* gene, int size);
void calc_fitness(gene_struct* gene_list, int gene_num, int size);
void make_ini_gene(gene_struct* gene_list, int gene_num, int size);
void p_to_o(gene_struct* gene_list,int gene_num,int size);
int index_num(list** num_list,int gene_el);
void o_to_p(gene_struct* gene_list,int gene_num,int size);
int pop_num(list** num_list,int gene_el);
void cross(gene_struct* gene_list,int gene_num, int size);
void mutation(gene_struct* gene_list, int gene_num, int size);
void gene_sort(gene_struct* gene_list, int gene_num, int size, int count);
void make_num_list(list* num_list, int size);

/* NQueen本体 */
int main(int argc, char const* argv[])
{
    start = clock();

    if(argc <= 1){
        puts("Use '[FILE NAME] [QUEEN NUM]'.");
        return 1;
    }
    
    srand(time(NULL));
    
    int size, gene_num, n;
    
    // Nの数
    size = strtol(argv[1], NULL, 10);
    
    //1世代あたりの遺伝子数
    gene_num = 10;
    
    //初期集団の生成
    gene_struct gene_list[gene_num];
    make_ini_gene(gene_list, gene_num, size);
    calc_fitness(gene_list, gene_num, size);
    /*
    printf("初期集団の生成\n");
    for(int i=0; i<gene_num; i++){
        for(int j=0; j<size; j++){
            printf("%d ", gene_list[i].gene[j]);
        }
        printf("  : %d \n", *gene_list[i].fitness);
    }
   */
    
    //学習ループ
    n = 1;
    
    //ループに入る
    while(1){
        //遺伝子が評価され，淘汰され，増殖する
        gene_sort(gene_list, gene_num, size, n);
        
        //交叉する
        p_to_o(gene_list,gene_num,size);
        cross(gene_list,gene_num, size);
        o_to_p(gene_list,gene_num,size);
        
        //突然変異する
        mutation(gene_list, gene_num, size);
        
        n++;
    }
    
    for(int i=0; i<gene_num; i++){       //領域の開放
        free(gene_list[i].gene);
        free(gene_list[i].fitness);
    }
    return 0;
}

/* Boardを出力 */
void board_print(int* gene, int size) {
    for(int i=0; i<size; i++){
        for(int j=0; j<size; j++){
            if(gene[i] == j)
                printf("Q ");
            else
                printf(". ");
        }
        printf("\n");
    }
}

/* 遺伝子生成 */
void make_ini_gene(gene_struct* gene_list, int gene_num, int size){
    for(int i=0; i<gene_num; i++){
        gene_list[i].gene = malloc(sizeof(int) * size); //遺伝子分の領域を確保
        gene_list[i].fitness = malloc(sizeof(int));
        for(int j=0; j<size; j++){
            gene_list[i].gene[j] = j;
        }
            
        for(int j=0; j<size; j++){              //遺伝子配列のシャッフル
            int n = rand() % size;
            int tmp = gene_list[i].gene[n];
            gene_list[i].gene[n] = gene_list[i].gene[j];
            gene_list[i].gene[j] = tmp;
        }
    }
}


/* 評価関数 */
void calc_fitness(gene_struct* gene_list, int gene_num, int size) {
    int g_vec[4][2] = {{-1, -1}, {-1, 1}, {1, -1}, {1, 1}};
    for(int i=0; i<gene_num; i++){
        *gene_list[i].fitness = 0;
        for(int j=0; j<size; j++){
            for(int vec=0; vec<4; vec++){
                for(int k=1; k<size; k++){
                    int x = g_vec[vec][1] * k + gene_list[i].gene[j];
                    int y = g_vec[vec][0] * k + j;
                    
                    if(x < 0 || x >= size || y < 0 || y >= size)
                        break;
                        
                    if(x == gene_list[i].gene[y])
                        *gene_list[i].fitness += 1;
                }
            }
        }
    }
}

/* 順列表現と順序表現へ変更 */
void p_to_o(gene_struct* gene_list,int gene_num,int size){
    list* num_list;
    
    //遺伝子変換
    for (int i = 0; i < gene_num; i++) {
        num_list = malloc(sizeof(list));
        make_num_list(num_list,size);
        for (int j = 0; j < size; j++) {
            gene_list[i].gene[j] = index_num(&num_list,gene_list[i].gene[j]);
        }
    }
   /* 
    printf("\n順列表現と順序表現へ変更\n");
    for(int i=0; i<gene_num; i++){
        for(int j=0; j<size; j++){
            printf("%d ", gene_list[i].gene[j]);
        }
        printf("  : %d \n", *gene_list[i].fitness);
    */
}

/* 要素のindexを返す */
int index_num(list** num_list, int gene_el) {
    list* TOP = *num_list;       //先頭アドレスを保持
    list* el_pre = NULL;        //参照している構造体の一つ前を指すポインタ変数
    int cnt = 0;                //カウンタ
    
    while(num_list != NULL){
        if((*num_list)->n == gene_el){
            if(el_pre == NULL)            //先頭に要素があった場合
                TOP = (*num_list)->next;
            else                            //それ以降に発見
                el_pre->next = (*num_list)->next;
            free(*num_list);
            *num_list = TOP;
            return cnt;
        }
        el_pre = *num_list;
        *num_list = (*num_list)->next;
        cnt++;
    }
    return -1;                  //見つからなかった場合，-1を返す
}

/* 順序表現から順列表現へ変換 */
void o_to_p(gene_struct* gene_list,int gene_num,int size) {
    list* num_list;
    
    //遺伝子変換
    for (int i = 0; i < gene_num; i++) {
        num_list = malloc(sizeof(list));
        make_num_list(num_list,size);
        for (int j = 0; j < size; j++) {
            gene_list[i].gene[j]  = pop_num(&num_list,gene_list[i].gene[j]);
        }
    }
   /* 
    printf("\n順序表現から順列表現へ変更\n");
    for(int i=0; i<gene_num; i++){
        for(int j=0; j<size; j++){
            printf("%d ", gene_list[i].gene[j]);
        }
        printf("  : %d \n", *gene_list[i].fitness);
    }*/
}

/* 要素をPOPする*/
int pop_num(list** num_list,int gene_el){
    list* TOP = *num_list;       //先頭アドレスを保持
    list* el_pre = NULL;        //参照している構造体の一つ前を指すポインタ変数
    int cnt = 0;                //カウンタ
    
    while(num_list != NULL){
        if(cnt == gene_el){
            if(el_pre == NULL)            //先頭に要素があった場合
                TOP = (*num_list)->next;
            else                            //それ以降に発見
                el_pre->next = (*num_list)->next;
            int num = (*num_list)->n;
            free(*num_list);
            *num_list = TOP;
            return num;
        }
        el_pre = *num_list;
        *num_list = (*num_list)->next;
        cnt++;
    }
    return -1;                  //見つからなかった場合，-1を返す
}


//num_listを生成
void make_num_list(list* num_list, int size){
    for(int i=0; i<size; i++){
        num_list->n = i;
        if(i != size-1){
            num_list->next = malloc(sizeof(list));
            num_list = num_list->next;
        }
        else
            num_list->next = NULL;
    }
}


/* 交叉*/
/* 遺伝子のリストを渡すと交叉する */
void cross(gene_struct* gene_list, int gene_num,  int size) {
    for(int i=0; i<gene_num/2; i++){
        int n = i * 2;
        int num = (rand() % size-3) + 1;
        int tmp[size];
       
        for(int j=0; j<size; j++){                  //n,n+1の配列を作成
            if(j < num)
                tmp[j] = gene_list[n].gene[j];
            else
                tmp[j] = gene_list[n+1].gene[j];
        } for(int j=0; j<size; j++){                  //n+1,nの配列を作成
            if(j >= num)
                gene_list[n+1].gene[j] = gene_list[n].gene[j];
        }
        
        for(int j=0; j<size; j++)
            gene_list[n].gene[j] = tmp[j];
    }
}

/* 突然変異 */
void mutation(gene_struct* gene_list, int gene_num, int size) {
    int gene, m, n, tmp;
    gene = rand() % gene_num;               //変異させる遺伝子を選択
    m = rand() % size;                      //入れ替える場所を２つ決める
    n = rand() % size;
    
    tmp = gene_list[gene].gene[m];
    gene_list[gene].gene[m] = gene_list[gene].gene[n];
    gene_list[gene].gene[n] = tmp;
}

/* 適応度を基準にソートし，淘汰と増殖を行う */
void gene_sort(gene_struct* gene_list, int gene_num, int size, int count){
    calc_fitness(gene_list, gene_num, size);
    
    //適応度をもとにソーティングを行う
    for(int i=1; i<gene_num; i++){
        int n = i;
        while(n > 0){
            if(*gene_list[n].fitness < *gene_list[n-1].fitness){
                gene_struct tmp = gene_list[n];
                gene_list[n] = gene_list[n-1];
                gene_list[n-1] = tmp;
            }
            n--;
        }
    }
    
    //最も良い適応度を出力
    //printf("%d\n", *gene_list[0].fitness);
    
    //適応度が0ならプログラムを終了させる
    if(*gene_list[0].fitness == 0){
        clock_t end = clock();
        printf("count : %d\n", count);
        board_print(gene_list[0].gene, size);
        printf("Time = %f\n", (double)(end-start)/(double)(1000000));
        exit(0);
    }
    
    for(int i=0; i<size; i++){
        gene_list[gene_num-1].gene[i] = gene_list[0].gene[i];
        gene_list[1].gene[i] = gene_list[0].gene[i];
    }
}
